### PR TITLE
Unfreeze PyMongo and replace deprecated count() with count_documents()

### LINKED ISSUE #121: unfreeze PyMongo
So the software will work with newer versions of MongoDB

### PR DESCRIPTION
**Pull request by northwestwitch**
### This PR adds | fixes:
- Unfreezed PyMongo in requirements.txt
- Replaced deprecated pymongo `.count()` function with `count_documents()` in code.
- Added basic cli tests touching fixed deprecated code

### How to test:
- Automatic tests should pass

### Expected outcome:
- [x] All tests should pass

### Review:
- [x] Code approved by HS
- [x] Tests executed by CR

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Review by henrikstranneheim — APPROVED** [review]
Looks good 👍 Test with a relevant CLI cmd perhaps?

**Comment by northwestwitch** [comment]
> Looks good 👍 Test with a relevant CLI cmd perhaps?

That's a good idea, I'll fix tomorrow. Thanks for the review!

**Comment by northwestwitch** [comment]
> Looks good 👍 Test with a relevant CLI cmd perhaps?

Mmm I've changed the automatic tests to run with Mongo 7 and a new version of PyMongo, but I see that there are some deprecated methods in the code that shouldn't be supported in the older versions of pymongo. Perhaps these are not covered by the tests. I'll fix!

**Comment by northwestwitch** [comment]
Ok, now I feel thsi ready for a review @henrikstranneheim, thanks!

File: loqusdb/commands/identity.py [hunk 1]
```diff
@@ -17,12 +17,11 @@ def identity(ctx, variant_id):
         ctx.abort()
 
     adapter = ctx.obj["adapter"]
-    version = ctx.obj["version"]
```
Comment (northwestwitch) [inline loqusdb/commands/identity.py:20]: Not used later in code

File: loqusdb/commands/export.py [hunk 2]
```diff
@@ -39,7 +39,6 @@ def export(ctx, outfile, variant_type, freq):
     version = ctx.obj["version"]
 
     LOG.info("Export the variants from {0}".format(adapter))
-    nr_cases = 0
```
Comment (northwestwitch) [inline loqusdb/commands/export.py:42]: Was not used later in code

File: loqusdb/commands/identity.py [hunk 3]
```diff
@@ -17,12 +17,11 @@ def identity(ctx, variant_id):
...
 
     LOG.info("Search variants {0}".format(adapter))
 
-    result = adapter.get_clusters(variant_id)
-    if result.count() == 0:
+    nr_results = adapter.db.identity.count_documents({"variant_id": variant_id})
```
Comment (henrikstranneheim) [inline loqusdb/commands/identity.py]:
```suggestion
    variant_count: int = adapter.db.identity.count_documents({"variant_id": variant_id})
```

File: loqusdb/commands/view.py [hunk 4]
```diff
@@ -42,10 +42,12 @@ def cases(ctx, case_id, to_json, count, case_type):
         case_obj["_id"] = str(case_obj["_id"])
         cases.append(case_obj)
     else:
-        cases = adapter.cases()
-        if cases.count() == 0:
+
+        nr_cases = adapter.case_count()
```
Comment (henrikstranneheim) [inline loqusdb/commands/view.py]:
```suggestion
        case_count: int = adapter.case_count()
```

File: loqusdb/commands/view.py [hunk 5]
```diff
@@ -42,10 +42,12 @@ def cases(ctx, case_id, to_json, count, case_type):
...
+        if nr_cases == 0:
             LOG.info("No cases found in database")
             ctx.abort()
+        cases = adapter.cases()
```
Comment (henrikstranneheim) [inline loqusdb/commands/view.py]: Add type hint

File: loqusdb/commands/identity.py [hunk 6]
```diff
@@ -17,12 +17,11 @@ def identity(ctx, variant_id):
...
+    if nr_results == 0:
         LOG.info("No hits for variant %s", variant_id)
```
Comment (henrikstranneheim) [inline loqusdb/commands/identity.py]: f-str

File: tests/commands/test_export.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+from click.testing import CliRunner
+
+from loqusdb.commands.cli import cli as base_command
+
+def test_export_base(real_db_name):
```
Comment (henrikstranneheim) [inline tests/commands/test_export.py]: Add type hint

File: tests/commands/test_export.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    """Test the base command that exports variants."""
+
+    runner = CliRunner()
+
+    # WHEN the base command to export cases is run
+    command = ["--database", real_db_name, "export"]
+
+    ## THEN it should return success
+    result = runner.invoke(base_command, command)
+    assert result.exit_code == 0
```
Comment (henrikstranneheim) [inline tests/commands/test_export.py:15]: Can we also test that it actually exported a file?

File: tests/commands/test_identity.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
+from click.testing import CliRunner
+
+from loqusdb.commands.cli import cli as base_command
+
+def test_identity(real_db_name):
```
Comment (henrikstranneheim) [inline tests/commands/test_identity.py]: Add type hint

File: tests/commands/test_view.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+from click.testing import CliRunner
+
+from loqusdb.commands.cli import cli as base_command
+
+def test_view_cases_base(real_mongo_adapter, real_db_name):
```
Comment (henrikstranneheim) [inline tests/commands/test_view.py]: Add type hints

File: tests/commands/test_view.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    ## GIVEN an empty database
+    assert sum([1 for _ in real_mongo_adapter.cases()]) == 0
+
+    runner = CliRunner()
+
+    # THEN the case command should return No cases found error
+    command = ["--database", real_db_name, "cases" ]
+    result = runner.invoke(base_command, command)
+    assert result.exit_code == 1
+    assert "No cases found in database" in result.output
```
Comment (henrikstranneheim) [inline tests/commands/test_view.py:17]:
```suggestion
    # WHEN counting cases in the database
    command = ["--database", real_db_name, "cases" ]
    result = runner.invoke(base_command, command)

    # THEN the case command should return No cases found error
    assert result.exit_code == 1
    assert "No cases found in database" in result.output
```

Comment (henrikstranneheim) [reply tests/commands/test_view.py:17]: Could also run with.pytest raises this error to test this properly

**Review by henrikstranneheim — APPROVED** [review]
Great that you added tests 👍

File: tests/commands/test_export.py (hunk 8 above)
Comment (northwestwitch) [reply tests/commands/test_export.py:15]: If you don't provide the -o param in the cli is not exporting to file, but printing into the terminal:

image

For the purpose of testing that the are 0 cases with the modified count command I think this is good enough. 
But I agree that more and better tests for the cli could have been written!

File: tests/commands/test_view.py (hunk 11 above)
Comment (northwestwitch) [reply tests/commands/test_view.py:17]: Mm it's not really raising an exception that you would catch in pytest with `with pytest.raises(ExceptionType) `(I tried)

not even `with pytest.raises(SystemExit)` works. 

It's a ctx.abort.() call I'm sure that there us a way to handle it using pytest, but I don't think it's worth the time spent googling for it. All the other cli tests are also written just like that, So I'd just leave it as it..

Comment (henrikstranneheim) [reply tests/commands/test_view.py:17]: 👍

**Pull request merged**